File: core/src/main/scala/com/graphbrain/eco/Prog.scala
package com.graphbrain.eco

import scala.io.Source
import com.graphbrain.eco.nodes.{WWRule, WVRule, ProgNode}

class Prog(val exprs: List[ProgNode]=List[ProgNode]()) {

  def wv(s: String, depth: Integer, caller: Context=null): List[Contexts] =
    wv(Words.fromString(s), depth, caller)

  def wv(w: Words, depth: Integer, caller: Context): List[Contexts] = {
    var ctxtsList = List[Contexts]()

    for (e <- exprs) e match {
      case wv: WVRule => {
        val ctxts = Contexts(wv, this, w, depth)
        wv.vertexValue(ctxts)
        if (ctxts.ctxts.size > 0) {
          ctxtsList ::= ctxts
          return ctxtsList
        }
      }
      case _ =>
    }

    ctxtsList
  }

  def ww(w: Words, depth: Integer, caller: Context): List[Contexts] = {
    var ctxtsList = List[Contexts]()

    for (e <- exprs) e match {
      case ww: WWRule => {
        val ctxts = Contexts(ww, this, w, depth)
        ww.wordsValue(ctxts)
        if (ctxts.ctxts.size > 0) {
          ctxtsList ::= ctxts
          return ctxtsList
        }
      }
      case _ =>
    }

    ctxtsList
  }

  def parse(sentence: String): String = {
    val ctxtList = wv(sentence, 0)

    var maxStrength = Double.NegativeInfinity
    for (ctxts <- ctxtList) {
      for (c <- ctxts.ctxts) {
        val strength = c.getNumber("_strength")
        if (strength > maxStrength) maxStrength = strength
      }
    }

    for (ctxts <- ctxtList) {
      for (c <- ctxts.ctxts) {
        val strength = c.getNumber("_strength")

        if (strength == maxStrength) {
          println(ctxts.sentence)
          return c.getTopRetVertex
        }
      }
    }

    ""
  }

  override def toString = exprs.map(_.toString).reduceLeft(_ + "\n" + _)
}

object Prog {
  private def emptyLine(line: String) =
    (line == "") || (line(0) == ';')

  def load(path: String) =
    fromStringList(Source.fromFile(path).getLines().toList)

  def fromString(str: String) =
    fromStringList(str.split("\\r?\\n").toList)

  def fromStringList(strList: List[String]) = {
    var exprStr = ""
    var exprList = List[ProgNode]()

    for(line <- strList) {
      if (emptyLine(line)) {
        if (exprStr != "") {
          val p = new Parser(exprStr)
          exprList ::= p.expr
        }
        exprStr = ""
      }
      else {
        exprStr += line
      }
    }

    if (exprStr != "") {
      val p = new Parser(exprStr)
      exprList ::= p.expr
    }

    new Prog(exprList.reverse)
  }

  def fromNode(e: ProgNode) = new Prog(List[ProgNode](e))

  def main(args: Array[String]) = {
    val p = Prog.load("eco/progs/test.eco")

    println(p)

    //val s = "Telmo likes chocolate."
    //val s = "Telmo likes eating chocolate."
    //val s = "The Obama administration is appealing to its allies in Congress."
    //val s = "The Obama administration is appealing to its allies in Congress, on Wall Street and across the country to stick with President Barack Obama's health care law even as embarrassing problems with the flagship website continue to mount."
    //val s = "The Obama administration is appealing to its allies in Congress to stick with health care law."
    //val s = "The research by America's Morgan Stanley financial services firm says demand for wine exceeded supply by 300m cases in 2012"
    //val s = "Egypt's ousted leader Mohammed Morsi has gone on trial in Cairo, telling the judge the case is illegitimate as he remains president."
    val s = "Egypt's ousted leader Mohammed Morsi has gone on trial in Cairo"

    val ctxtList = p.wv(s, 0)

    /*
    var maxStrength = Double.NegativeInfinity
    for (ctxts <- ctxtList) {
      for (c <- ctxts.ctxts) {
        val strength = c.getNumber("_strength")
        if (strength > maxStrength) maxStrength = strength
      }
    }
    */

    for (ctxts <- ctxtList) {
      println(ctxts.sentence)

      for (c <- ctxts.ctxts) {
        val strength = c.getNumber("_strength")

        //if (strength == maxStrength) {
          println("\n\n")
          println(c.getTopRetVertex)
          println(strength)
          c.printCallStack()
        //}
      }
    }
  }
}

File: core/src/main/java/com/graphbrain/eco/Prog.java
package com.graphbrain.eco;

import com.graphbrain.eco.nodes.WWRule;
import com.graphbrain.eco.nodes.WVRule;
import com.graphbrain.eco.nodes.ProgNode;
import com.graphbrain.db.Vertex;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

public class Prog {

    private List<ProgNode> exprs;

    public Prog(List<ProgNode> exprs) {
        this.exprs = exprs;
    }

    public Prog() {
        this(new LinkedList<ProgNode>());
    }

    public List<Contexts> wv(String s, int depth) {
        return wv(s, depth, null);
    }

    public List<Contexts> wv(String s, int depth, Context globals) {
        return wv(Words.fromString(s), depth, globals);
    }

    public List<Contexts> wv(Words w, int depth, Context globals) {
        List<Contexts> ctxtsList = new LinkedList<>();

        for (ProgNode e : exprs) {
            if (e instanceof WVRule) {
                WVRule wv = (WVRule)e;
                Contexts ctxts = new Contexts(wv, this, w, depth);
                ctxts.setGlobals(globals);
                wv.eval(ctxts);
                if (ctxts.getCtxts().size() > 0) {
                    ctxtsList.add(ctxts);
                    return ctxtsList;
                }
            }
        }

        return ctxtsList;
    }

    public List<Contexts> ww(Words w, int depth, Context globals) {
        List<Contexts> ctxtsList = new LinkedList<>();

        for (ProgNode e : exprs) {
            if (e instanceof WWRule) {
                WWRule ww = (WWRule)e;
                Contexts ctxts = new Contexts(ww, this, w, depth);
                ctxts.setGlobals(globals);
                ww.eval(ctxts);
                if (ctxts.getCtxts().size() > 0) {
                    ctxtsList.add(ctxts);
                    return ctxtsList;
                }
            }
        }

        return ctxtsList;
    }

    public Vertex parse(String sentence) {
        List<Contexts> ctxtList = wv(sentence, 0);

        double maxStrength = Double.NEGATIVE_INFINITY;
        for (Contexts ctxts : ctxtList) {
            for (Context c : ctxts.getCtxts()) {
                double strength = c.getNumber("_strength");
                if (strength > maxStrength)
                    maxStrength = strength;
            }
        }

        for (Contexts ctxts : ctxtList) {
            for (Context c : ctxts.getCtxts()) {
                double strength = c.getNumber("_strength");

                if (strength == maxStrength) {
                    //System.out.println(ctxts.sentence);
                    return c.getTopRetVertex();
                }
            }
        }

        return null;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(100);
        for (ProgNode e : exprs) {
            sb.append(e.toString());
            sb.append('\n');
        }
        return sb.toString();
    }

    private static boolean emptyLine(String line) {
        return ((line.equals("")) || (line.charAt(0) == ';'));
    }

    public static Prog load(String path) {
        try {
            String str = FileUtils.readFileToString(new File(path));
            return fromString(str);
        }
        catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static Prog fromString(String str) {
        return fromStringArray(str.split("\\r?\\n"));
    }

    public static Prog fromStringArray(String[] strArr) {
        String exprStr = "";
        List<ProgNode> exprList = new LinkedList<>();

        for(String line : strArr) {
            if (emptyLine(line)) {
                if (!exprStr.isEmpty()) {
                    Parser p = new Parser(exprStr);
                    exprList.add(p.getExpr());
                }
                exprStr = "";
            }
            else {
                exprStr += line;
            }
        }

        if (!exprStr.isEmpty()) {
            Parser p = new Parser(exprStr);
            exprList.add(p.getExpr());
        }

        return new Prog(exprList);
    }

    public static Prog fromNode(ProgNode e) {
        List<ProgNode> l = new LinkedList<>();
        l.add(e);
        return new Prog(l);
    }
}